### DISCUSSION TITLE
Element Sources registry

### DISCUSSION THREAD

**Discussion by brandonkelly** [craftcms/cms#8422] [Ideas] [answered]
If you have an asset volume with a ton of subfolders on the same level, JS has to initialize them all at once when they are first made visible on the Assets index page, which can take a while.

To fix this, there should be an Element Sources Registry, which contains a full record of all element types’ sources, and which could be loaded in stages. So when expanding an asset volume, for example, its subfolders could be lazy loaded in batches of 50 or 100.

This would probably manifest as an `ElementSources` service with a few methods that element types could use to keep their source lists updated. Behind the scenes, the sources would be stored in an `elementsources` DB table. Element types’ `getSources()` functions would go away.

**Comment by sjelfull** [comment]
Some UX feedback on this:
It should be very quick to jump to anywhere in the source list. 

As an example: Say i'm looking for a folder named "Images from Summer School". I don't want to scroll down 50 or 100 at a time, wait for the loading to finish, scroll some more, until I get to the folder.

Likewise, it should be possible to get back to a position in a element source. In this case, that can be a query parameter + a A-Z jump nav. Related to #1596 and #818

**Comment by leevigraham** [comment]
Adding the UX feedback… ordering numbered folders should use numeric, not text ordering:

<img width="249" alt="image" src="https://user-images.githubusercontent.com/25124/80442840-28516280-8951-11ea-867d-ca4790bb4f97.png">

**Comment by brandonkelly** [comment, marked as answer]
This idea is no longer being considered, in favor of #9171, which will come in Craft 3.8 and 4.4. (See the [Rethinking Volume Subfolders](https://craftcms.com/blog/rethinking-volume-subfolders) blog post and PR #12558 for more info.)

**Reply by brandonkelly** [reply]
Craft 3.8 and 4.4 are both out now, with those changes 🎉
